simple_certmanager/utils.py

```diff
@@ -1,67 +1,82 @@
-    end = pem.find(mark) + len(mark)
-    yield pem[:end]
-    yield from split_pem(pem[end:])
+    Nothing in this code is really specific to x509, but the function name expresses
+    our *intent* to only deal with x509 certificate/key pairs.
+    """
+    # passwords are not supported
+    return load_pem_private_key(data, password=None)
 
 
-def load_pem_chain(pem: bytes) -> Generator[x509.Certificate, None, None]:
-    for data in split_pem(pem):
-        yield x509.load_pem_x509_certificate(data)
+def pretty_print_certificate_components(x509name: x509.Name) -> str:
+    bits = (f"{attr.rfc4514_attribute_name}: {attr.value}" for attr in x509name)
+    return ", ".join(bits)
 
 
-def check_pem(
-    pem: bytes,
-    ca: Union[bytes, str, PathLike] = certifi.where(),
-    ca_path: Optional[Union[str, PathLike]] = None,
-) -> bool:
+def check_pem(pem: bytes, ca: str | Path = certifi.where()) -> bool:
     """Simple (possibly incomplete) sanity check on pem chain.
 
     If the pam passes this check it MAY be valid for use. This is only intended
     to catch blatant misconfigurations early. This gives NO guarantees on
     security nor authenticity.
-    """
-    # We need still need to use pyOpenSSL primitives for this:
-    # https://github.com/pyca/cryptography/issues/6229
-    # https://github.com/pyca/cryptography/issues/2381
 
-    # Establish roots
-    store = crypto.X509Store()
-    store.load_locations(ca, ca_path)
+    Relevant documentation: https://cryptography.io/en/stable/x509/verification/
+
+    See all the context in the upstream issue:
+    https://github.com/pyca/cryptography/issues/2381
 
-    leaf, *chain = map(crypto.X509.from_cryptography, load_pem_chain(pem))
+    :arg pem: a certificate or chain of certificates in PEM format.
+    :arg ca: path to the root Certificate Authority bundle.
+
+    .. todo: The default for ``ca`` should probably support a setting so that
+       self_certifi paths/dirs can be taken into account, or maybe consider the envvar
+       ``REQUESTS_CA_BUNDLE``. This will make it possible to support the G1 Private root.
+    """
+    # normalize to Path
+    if isinstance(ca, str):
+        ca = Path(ca)
+
+    [leaf, *intermediates] = x509.load_pem_x509_certificates(pem)
+    root_certificates = x509.load_pem_x509_certificates(ca.read_bytes())
+
+    store = Store(root_certificates)
+    builder = PolicyBuilder().store(store)
+
+    # extract the DNS name from the leaf certificate - we don't really care about the
+    # exact host name for the chain validation since we don't know which hosts will be
+    # connected to with this certificate - that only happens at runtime. So, we use the
+    # first entry.
+    # XXX: this can crash if the SAN extension is not present or no DNS names are
+    # specified - it's unclear if those are situations we need to support or not.
+    # Probably the validation of the (leaf) certificate itself should enforce the
+    # presence of this information.
```

Comment:
Two admin tests are now failing on this, I'm not sure how to handle this yet :thinking: